### PR TITLE
Idempotency enclosure group

### PR DESCRIPTION
**Pull request by nabhajit-ray**
### Description
[Describe what this change achieves]

### Issues Resolved
[List any issues this PR will resolve. e.g., Fixes #01]

### Check List
- [ ] New functionality includes testing.
  - [ ]  All tests pass for go 1.11 + gofmt checks.
- [ ] New functionality has been documented in the README if applicable.
  - [ ] New functionality has been thoroughly documented in the examples (please include helpful comments).
  - [ ] New endpoints supported are updated in the endpoints-support.md file.
- [ ] Changes are documented in the CHANGELOG.

### CODE REVIEW

File: oneview/resource_enclosure_group.go [hunk 1]
```diff
@@ -46,10 +47,11 @@ func resourceEnclosureGroup() *schema.Resource {
 			"description": {
 				Type:     schema.TypeString,
 				Optional: true,
+				Computed: true,
```
Comment (AsisBagga) [inline oneview/resource_enclosure_group.go:50]: Is this getting auto-populated?
Setting Description to computed+optional means you cant set it to null in the update operation. 
I saw a use case scenario in REST Guide, so raised this comment.

File: oneview/resource_logical_enclosure.go [hunk 2]
```diff
@@ -82,15 +83,18 @@ func resourceLogicalEnclosure() *schema.Resource {
 
 			"enclosure_group_uri": {
 				Type:     schema.TypeString,
-				Required: true,
+				Computed: true,
+				Optional: true,
 			},
 			"enclosure_uris": {
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:89]: This is a read only field. Therefore, we can remove optional.
And if oneview is not sending an ordered array, then making this attribute to TypeSet would be a good idea.

File: oneview/resource_logical_enclosure.go [hunk 3]
```diff
@@ -82,15 +83,18 @@ func resourceLogicalEnclosure() *schema.Resource {
...
-				Required: true,
+				Computed: true,
+				Optional: true,
 				Type:     schema.TypeList,
 				Elem:     &schema.Schema{Type: schema.TypeString},
 			},
 			"firmware": {
 				Optional: true,
+				Computed: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:97]: why have we set firmware to computed when all the attributes inside it are optional.

File: oneview/resource_logical_enclosure.go [hunk 4]
```diff
@@ -160,6 +165,7 @@ func resourceLogicalEnclosure() *schema.Resource {
 			},
 			"logical_interconnect_uris": {
 				Optional: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:167]: This is a read only field please remove optional from here.

File: oneview/resource_logical_enclosure.go [hunk 5]
```diff
@@ -171,10 +177,12 @@ func resourceLogicalEnclosure() *schema.Resource {
 			"power_mode": {
 				Type:     schema.TypeString,
 				Optional: true,
+				Computed: true,
 			},
 			"scaling_state": {
 				Type:     schema.TypeString,
 				Optional: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:184]: This is a read only field please remove optional from here.

File: oneview/resource_logical_enclosure.go [hunk 6]
```diff
@@ -183,14 +191,17 @@ func resourceLogicalEnclosure() *schema.Resource {
 			"status": {
 				Type:     schema.TypeString,
 				Optional: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:193]: This is a read only field please remove optional from here.

File: oneview/resource_logical_enclosure.go [hunk 7]
```diff
@@ -183,14 +191,17 @@ func resourceLogicalEnclosure() *schema.Resource {
...
+				Computed: true,
 			},
 			"type": {
 				Type:     schema.TypeString,
 				Optional: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:198]: This is a read only field please remove optional from here.

File: oneview/resource_logical_enclosure.go [hunk 8]
```diff
@@ -183,14 +191,17 @@ func resourceLogicalEnclosure() *schema.Resource {
...
+				Computed: true,
 			},
 			"uri": {
 				Type:     schema.TypeString,
 				Optional: true,
```
Comment (AsisBagga) [inline oneview/resource_logical_enclosure.go:203]: This is a read only field please remove optional from here.

File: oneview/resource_logical_enclosure.go (hunk 3 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:97]: done

File: oneview/resource_logical_enclosure.go (hunk 2 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:89]: done

File: oneview/resource_logical_enclosure.go (hunk 4 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:167]: done

File: oneview/resource_logical_enclosure.go (hunk 5 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:184]: done

File: oneview/resource_logical_enclosure.go (hunk 6 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:193]: done

File: oneview/resource_logical_enclosure.go (hunk 7 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:198]: done

File: oneview/resource_ethernet_network.go [hunk 9]
```diff
@@ -109,6 +105,7 @@ func resourceEthernetNetwork() *schema.Resource {
 			},
 			"subnet_uri": {
 				Optional: true,
+				Computed: true,
 				Type:     schema.TypeString,
 			},
 			"initial_scope_uris": {
```
Comment (AvijitOS) [inline oneview/resource_ethernet_network.go:111]: same comments as abover

File: oneview/resource_logical_enclosure.go [hunk 10]
```diff
@@ -171,26 +173,28 @@ func resourceLogicalEnclosure() *schema.Resource {
...
+				Computed: true,
 			},
 			"scaling_state": {
 				Type:     schema.TypeString,
-				Optional: true,
+				Computed: true,
 			},
 			"scopes_uri": {
 				Type:     schema.TypeString,
 				Optional: true,
```
Comment (AvijitOS) [inline oneview/resource_logical_enclosure.go:184]: can we change scopes_uri once element created?

File: oneview/resource_fc_network.go [hunk 11]
```diff
@@ -164,8 +158,13 @@ func resourceFCNetworkRead(d *schema.ResourceData, meta interface{}) error {
...
 	d.Set("etag", fcNet.ETAG)
-	d.Set("scopeuri", fcNet.ScopesUri.String())
-	d.Set("initial_scope_uris", fcNet.InitialScopeUris)
+	d.Set("scopesuri", fcNet.ScopesUri.String())
+	//d.Set("initial_scope_uris", fcNet.InitialScopeUris)
+	initialscopeuris := make([]string, 0, len(fcNet.InitialScopeUris))
+	for _, initialscopeuri := range fcNet.InitialScopeUris {
+		initialscopeuris = append(initialscopeuris, initialscopeuri.String())
+	}
+	d.Set("initial_scope_uris", initialscopeuris)
```
Comment (AvijitOS) [inline oneview/resource_fc_network.go:167]: during get we don't get initial_scope_uri so is it necessary

File: oneview/resource_enclosure_group.go [hunk 12]
```diff
@@ -69,22 +67,18 @@ func resourceEnclosureGroup() *schema.Resource {
 			},
 			"interconnect_bay_mapping_count": {
 				Type:     schema.TypeInt,
-				Optional: true,
+				Computed: true,
 			},
 			"interconnect_bay_mappings": {
```
[the remark this answers is not in the document]
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go:72]: description is there in the rest api doc, and it is part of PUT call also. So I belve we can keep it optional.

[the remark this answers is not in the document]
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go:72]: Empty bay mapping works

File: oneview/resource_fc_network.go (hunk 11 above)
Comment (nabhajit-ray) [reply oneview/resource_fc_network.go:167]: done

File: examples/ethernet_networks/update_resource.tf [hunk 13]
```diff
@@ -10,17 +10,20 @@ provider "oneview" {
 # Updates the resource created from main.tf
 resource "oneview_ethernet_network" "ethernetnetwork" {
   name    = "TestEthNetwork_terraform_Rename"
+  ethernet_network_type = "Tagged"
   type    = "ethernet-networkV4"
   vlan_id = 100
 }
 resource "oneview_ethernet_network" "ethernetnetwork_1" {
   name    = "Auto-Ethernet-1"
+  ethernet_network_type = "Tagged"
```
Comment (AvijitOS) [inline examples/ethernet_networks/update_resource.tf:19]: can we have example dealing with different ethernet networks like untagged, tunnel

File: examples/fc_networks/main.tf [hunk 14]
```diff
@@ -14,6 +14,7 @@ data "oneview_scope" "scope_obj" {
 # Creates the fc network with name TestFCNetwork and local name FCNetwork.
 resource "oneview_fc_network" "FCNetwork" {
   name               = "TestFCNetwork"
+  fabric_type        = "FabricAttach"
```
Comment (AvijitOS) [inline examples/fc_networks/main.tf:17]: its better to provide examples for fabric_type direct attach

File: oneview/resource_logical_enclosure.go (hunk 10 above)
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:184]: No scope uri can not be changed

File: oneview/resource_enclosure_group.go [hunk 15]
```diff
@@ -40,24 +41,21 @@ func resourceEnclosureGroup() *schema.Resource {
...
 				Type:     schema.TypeInt,
 				Optional: true,
+				Default:  1,
 			},
 			"enclosure_type_uri": {
 				Type:     schema.TypeString,
 				Optional: true,
+				Computed: true,
 			},
 			"initial_scope_uris": {
```
[the remark this answers is not in the document]
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go:60]: Done

File: examples/ethernet_networks/update_resource.tf (hunk 13 above)
Comment (nabhajit-ray) [reply examples/ethernet_networks/update_resource.tf:19]: Done

File: examples/fc_networks/main.tf (hunk 14 above)
Comment (nabhajit-ray) [reply examples/fc_networks/main.tf:17]: Done

File: examples/ethernet_networks/main.tf [hunk 16]
```diff
@@ -9,20 +9,38 @@ provider "oneview" {
...
-  name    = "Auto-Ethernet-2"
-  type    = "ethernet-networkV4"
-  vlan_id = 102
+  name                  = "Auto-Ethernet-2"
+  ethernet_network_type = "Tagged"
+  type                  = "ethernet-networkV4"
+  vlan_id               = 102
 }
 
+#Creating tunnel ethernet network
```
Comment (AsisBagga) [inline examples/ethernet_networks/main.tf:32]: can you remove this if not required

File: examples/ethernet_networks/main.tf [hunk 17]
```diff
@@ -9,20 +9,38 @@ provider "oneview" {
...
+# resource "oneview_ethernet_network" "ethernetnetwork_2" {
+#   name                  = "Auto-Ethernet-2"
+#   ethernet_network_type = "Tunnel"
+#   type                  = "ethernet-networkV4"
+
+# }
+
+# resource "oneview_ethernet_network" "ethernetnetwork_2" {
```
Comment (AsisBagga) [inline examples/ethernet_networks/main.tf:40]: can you remove this if not required

File: examples/fc_networks/main.tf [hunk 18]
```diff
@@ -14,7 +14,15 @@ data "oneview_scope" "scope_obj" {
...
   type               = "fc-networkV4"
   initial_scope_uris = [data.oneview_scope.scope_obj.uri]
 }
+# Creates the fc network with  fabric type direct attach.
```
Comment (AsisBagga) [inline examples/fc_networks/main.tf:21]: can you remove this if not required

File: oneview/resource_enclosure_group.go [hunk 19]
```diff
@@ -191,18 +201,24 @@ func resourceEnclosureGroupCreate(d *schema.ResourceData, meta interface{}) erro
...
+	interconnectBayMappings := make([]ov.InterconnectBayMap, 0)
 	if val, ok := d.GetOk("interconnect_bay_mappings"); ok {
 		rawInterconnectBayMappings := val.(*schema.Set).List()
-		interconnectBayMappings := make([]ov.InterconnectBayMap, 0)
 		for _, raw := range rawInterconnectBayMappings {
 			interconnectBayMappingItem := raw.(map[string]interface{})
+			logicalInterconnectGroup, err := config.ovClient.GetLogicalInterconnectGroupByName(interconnectBayMappingItem["logical_interconnect_group_name"].(string))
+			if err != nil || logicalInterconnectGroup.URI.IsNil() {
+				d.SetId("")
+				return nil
```
Comment (AsisBagga) [inline oneview/resource_enclosure_group.go:212]: can we return an error statement here please?

File: oneview/resource_enclosure_group.go [hunk 20]
```diff
@@ -279,22 +304,41 @@ func resourceEnclosureGroupRead(d *schema.ResourceData, meta interface{}) error
...
 	d.Set("name", enclosureGroup.Name)
 
+	interconnectBayMap := make([]map[string]interface{}, 0, 1)
+	for i := 0; i < len(enclosureGroup.InterconnectBayMappings); i++ {
+		liguri := enclosureGroup.InterconnectBayMappings[i].LogicalInterconnectGroupUri
+
+		encLIG, err := config.ovClient.GetLogicalInterconnectGroupByUri(liguri)
+		if err != nil || encLIG.Name == "" {
+			d.SetId("")
+			return nil
```
Comment (AsisBagga) [inline oneview/resource_enclosure_group.go:325]: please return an error statementif no LIG is found.

File: oneview/resource_enclosure_group.go [hunk 21]
```diff
@@ -330,11 +374,15 @@ func resourceEnclosureGroupUpdate(d *schema.ResourceData, meta interface{}) erro
 	interconnectBayMappings := make([]ov.InterconnectBayMap, 0)
 	for _, raw := range rawInterconnectBayMappings {
 		interconnectBayMappingItem := raw.(map[string]interface{})
-
+		logicalInterconnectGroup, err := config.ovClient.GetLogicalInterconnectGroupByName(interconnectBayMappingItem["logical_interconnect_group_name"].(string))
+		if err != nil || logicalInterconnectGroup.URI.IsNil() {
+			d.SetId("")
```
Comment (AsisBagga) [inline oneview/resource_enclosure_group.go:379]: same as above.

File: oneview/resource_fcoe_network.go [hunk 22]
```diff
@@ -54,10 +55,6 @@ func resourceFCoENetwork() *schema.Resource {
 				Type:     schema.TypeString,
 				Computed: true,
 			},
-			"description": {
-				Type:     schema.TypeString,
-				Optional: true,
-			},
```
Comment (AvijitOS) [inline oneview/resource_fcoe_network.go:60]: scopes_uri also be read only attribute..so remove optional part from it

File: oneview/resource_fc_network.go [hunk 23]
```diff
@@ -34,9 +36,7 @@ func resourceFCNetwork() *schema.Resource {
 			},
 			"fabric_type": {
 				Type:     schema.TypeString,
-				Optional: true,
-				ForceNew: true,
-				Default:  "FabricAttach",
+				Required: true,
 			},
 			"link_stability_time": {
```
Comment (AvijitOS) [inline oneview/resource_fc_network.go:41]: remove optional from scopes_uri, it should be only computed

File: oneview/resource_enclosure_group.go [hunk 24]
```diff
@@ -66,25 +64,25 @@ func resourceEnclosureGroup() *schema.Resource {
 					Type: schema.TypeString,
 				},
 				Set: schema.HashString,
```
Comment (AvijitOS) [inline oneview/resource_enclosure_group.go:66]: as discussed we should handle it differently

File: oneview/resource_enclosure_group.go (hunk 19 above)
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go]: Done

File: oneview/resource_enclosure_group.go (hunk 20 above)
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go]: Done

File: oneview/resource_enclosure_group.go (hunk 24 above)
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go]: Done

File: oneview/resource_enclosure_group.go (hunk 21 above)
Comment (nabhajit-ray) [reply oneview/resource_enclosure_group.go:378]: Done

File: oneview/resource_logical_enclosure.go [hunk 25]
```diff
@@ -392,15 +399,7 @@ func resourceLogicalEnclosureUpdate(d *schema.ResourceData, meta interface{}) er
...
-	enclosureSetCount := d.Get("enclosure_uris.#").(int)
-	enclosureUris := make([]utils.Nstring, enclosureSetCount)
-	for i := 0; i < enclosureSetCount; i++ {
-		enclosureSetPrefix := fmt.Sprintf("enclosure_uris.%d", i)
-		if val, ok := d.GetOk(enclosureSetPrefix); ok {
-			enclosureUris[i] = utils.NewNstring(val.(string))
-		}
-	}
-	logicalEnclosure.EnclosureUris = enclosureUris
+
```
[the remark this answers is not in the document]
Comment (nabhajit-ray) [reply oneview/resource_logical_enclosure.go:402]: Done

**Review by AvijitOS — APPROVED** [review]

**Review by AsisBagga — APPROVED** [review]
LGTM

**Pull request merged**